kore/src/Kore/Unification/Substitution.hs

```diff
@@ -219,6 +226,98 @@ partition criterion (NormalizedSubstitution substitution) =
     let (true, false) = Map.partitionWithKey criterion substitution
     in (NormalizedSubstitution true, NormalizedSubstitution false)
 
+{- | Reverses all `var = givenVar` assignments in the substitution and
+renormalizes, if needed.
+-}
+reverseIfRhsIsVar
+    :: forall variable
+    .   ( Eq variable
+        , FreshVariable variable
+        , Show variable
+        , SortedVariable variable
+        , Unparse variable
+        )
+    => variable
+    -> Substitution variable
+    -> Substitution variable
+reverseIfRhsIsVar variable (Substitution substitution) =
+    Substitution (map (reversePairIfRhsVar variable) substitution)
+  where
+    reversePairIfRhsVar
+        :: variable
+        -> (variable, TermLike variable)
+        -> (variable, TermLike variable)
+    reversePairIfRhsVar var original@(substitutedVar, Var_ substitutionVar)
+      | var == substitutionVar = (substitutionVar, mkVar substitutedVar)
+      | otherwise = original
+    reversePairIfRhsVar _ original = original
+reverseIfRhsIsVar variable original@(NormalizedSubstitution substitution) =
+    case reversableVars of
+        [] -> original
+        (v:vs) ->
+            let
+                replacementVar :: variable
+                replacementVar = foldr max v vs
```

Comment:
We should probably use `foldl'` here because the result is a summary of the input and it will be immediately used.